pkg/virtctl/create/instancetype/instancetype_test.go

```diff
@@ -0,0 +1,183 @@
+			Expect(*instancetypeSpec.IOThreadsPolicy).To(Equal(policy))
+
+		},
+			Entry("VirtualMachineInstacetype set to auto", namespaced, "auto", v1.IOThreadsPolicyAuto),
+			Entry("VirtualMachineInstacetype set to shared", namespaced, "shared", v1.IOThreadsPolicyShared),
+
+			Entry("VirtualMachineClusterInstacetype set to auto", "", "auto", v1.IOThreadsPolicyAuto),
+			Entry("VirtualMachineClusterInstacetype set to shared", "", "shared", v1.IOThreadsPolicyShared),
+		)
+
+		DescribeTable("Invalid arguments", func(namespaced, flag, params, errMsg string) {
+			cmd := clientcmd.NewRepeatableVirtctlCommand(create, Instancetype, namespaced,
+				setFlag(CPUFlag, "1"),
+				setFlag(MemoryFlag, "128Mi"),
+				setFlag(flag, params),
+			)
+			err := cmd()
+
+			Expect(err).To(HaveOccurred())
+			Expect(err).To(MatchError(errMsg))
+		},
+			Entry("VirtualMachineInstacetype gpu missing name", namespaced, GPUFlag, "devicename:nvidia", fmt.Sprintf("failed to parse \"--gpu\" flag: %+s", NameErr)),
+			Entry("VirtualMachineInstacetype gpu missing deviceName", namespaced, GPUFlag, "name:gpu1", fmt.Sprintf("failed to parse \"--gpu\" flag: %+s", DeviceNameErr)),
+			Entry("VirtualMachineInstacetype hostdevice missing name", namespaced, HostDeviceFlag, "devicename:intel", fmt.Sprintf("failed to parse \"--hostdevice\" flag: %+s", NameErr)),
+			Entry("VirtualMachineInstacetype hostdevice missing deviceName", namespaced, HostDeviceFlag, "name:device1", fmt.Sprintf("failed to parse \"--hostdevice\" flag: %+s", DeviceNameErr)),
+			Entry("VirtualMachineInstacetype to IOThreadsPolicy", namespaced, IOThreadsPolicyFlag, "invalid-policy", fmt.Sprintf("failed to parse \"--iothreadspolicy\" flag: %+s", IOThreadErr)),
+
+			Entry("VirtualMachineClusterInstacetype gpu missing name", "", GPUFlag, "devicename:nvidia", fmt.Sprintf("failed to parse \"--gpu\" flag: %+s", NameErr)),
+			Entry("VirtualMachineClusterInstacetype gpu missing deviceName", "", GPUFlag, "name:gpu1", fmt.Sprintf("failed to parse \"--gpu\" flag: %+s", DeviceNameErr)),
+			Entry("VirtualMachineClusterInstacetype hostdevice missing name", "", HostDeviceFlag, "devicename:intel", fmt.Sprintf("failed to parse \"--hostdevice\" flag: %+s", NameErr)),
+			Entry("VirtualMachineClusterInstacetype hostdevice missing deviceName", "", HostDeviceFlag, "name:device1", fmt.Sprintf("failed to parse \"--hostdevice\" flag: %+s", DeviceNameErr)),
+			Entry("VirtualMachineClusterInstacetype to IOThreadsPolicy", "", IOThreadsPolicyFlag, "invalid-policy", fmt.Sprintf("failed to parse \"--iothreadspolicy\" flag: %+s", IOThreadErr)),
+		)
+	})
+})
+
+func setFlag(flag, parameter string) string {
+	return fmt.Sprintf("--%s=%s", flag, parameter)
+}
+
+func runCmd(args ...string) ([]byte, error) {
+	_args := append([]string{create}, args...)
+	return clientcmd.NewRepeatableVirtctlCommandWithOut(_args...)()
+}
+
+func getInstancetypeSpec(namespaced string, bytes []byte) *instancetypev1alpha2.VirtualMachineInstancetypeSpec {
```

Comment:
Could be done more elegantly with a generic func?

Reply:
You can't access common attributes shared by both types so I'm not sure it can?

Reply:
I'm always getting fooled by it. Maybe with an `interface{}` and switch/case type cast instead.

Reply:
I just don't like passing in a string and checking for its length to determine what to do.

Reply:
We need to decide here what type should be unmarshaled from the output, which is some bytes, and the only way to know if it is `instancetype` or `clusterInstancetype` is to get the information from  command input parameters. I am not sure if it could be done with the `interface{}` here. I was trying to use `interface{}` and `switch/case` in a instancetype.go, but did not manage to make it work.

Reply:
> I just don't like passing in a string and checking for its length to determine what to do.

Yeah a `bool` would be cleaner IMHO but this is just unit tests so meh.

> We need to decide here what type should be unmarshaled from the output, which is some bytes, and the only way to know if it is `instancetype` or `clusterInstancetype` is to get the information from command input parameters. I am not sure if it could be done with the `interface{}` here. I was trying to use `interface{}` and `switch/case` in a instancetype.go, but did not manage to make it work.

Would something like the following work if you unmarshal into a runtime.object first?

https://github.com/kubevirt/kubevirt/blob/2d470d8aa2f37df9ca5695d4b347ff8d87011c6d/pkg/instancetype/instancetype.go#L436-L443

Reply:
To answer my own question, yes the following works for me:

https://gist.github.com/4588d3be2baec7fb6cdcfba00c60810f